### DISCUSSION TITLE
Json formatting isn't working

### DISCUSSION THREAD

**Discussion by sgatu** [rest-nvim/rest.nvim#457] [Q&A]
Hi, I've just installed rest.nvim in my setup and it looks pretty promising, but can't figure out why it is not formatting the JSON output.  I'm pretty sure the API answers with a Content-Type: application/json.
![restnvim](https://github.com/user-attachments/assets/5df073a1-a90d-4553-9f01-b1753b8f2f34)

I'm using LazyVim with some basic configuration:

```
return {
  {
    "rest-nvim/rest.nvim",
    config = function() end,
    opts = {
      highlight = {
        enabled = true,
        timeout = 150,
      },
    },
  },
} 
```
~~And I have installed gq and I can use it from my terminal.~~
Thought that I needed to install gq, but gq is referring to a vim motion. 

Also this in a bit on the tangent, most of configuration examples I find online are totally different from the README.md example, I assume that the library has changed and that's why, but the result split configurations looked pretty nice and I'm not sure there are available anymore.

**Comment by sgatu** [comment]
For now I've forked the project and changed 
`vim.cmd("silent normal gggqG")`
with
`vim.cmd("%!jq")`

If I execute the vim motion gggqG on a separate file it formats well, but the output of the rest.nvim is not formatted. 

Again, not sure if the problem is a misconfiguration or that there is a bug in the library.

**Reply by sgatu** [reply]
At this point I'm getting a bit mad, tried everything and can't understand why it does not work with the motion. 

Looks like the buffer filetype is correct set to JSON, there is no error in the command `silent normal gggqG` and if I try to open a json file from the project and I do the motion `gggqG` it formats correctly.

Not only that, if I execute `lua vim.cmd("silent normal gggqG")` on the opened JSON file it also works as intended,  yet the output of the request is not getting formatted for some strange reason.

**Comment by jrebs** [comment]
Create an ftplugin file (I used `$NVIM/fplugin/json.lua`) with the following contents:

```
vim.bo.formatexpr = “”
vim.bo.formatprg = “jq”
```
Source: https://github.com/rest-nvim/rest.nvim/issues/414

**Reply by xaviduds** [reply]
Hello, @jrebs , I read #414 to the end, I just couldn't find what file to type that into:

```bash
$NVIM/fplugin/json.lua
bash: /fplugin/json.lua: No such file or directory
```

I send a photo of my my config for rest.nvim, the usage with GET at blabla.http, the result, with unformatted json, I use none-ls with prettier correctly configured so a random a.json file (bottom-right) is correctly formatted (I copy-pasted the result from rest unformated json):

![screenshot](https://github.com/user-attachments/assets/d2fa3634-ed93-4ab3-869c-af11b7188374)

i tried putting this in my options.lua:
```lua
vim.bo.formatexpr = ""
vim.bo.formatprg = "jq"
```

I also have jq installed, and doing \<leader\>gq does nothing to me

**Reply by jrebs** [reply]
$NVIM is obviously a environment variable. In your output you pasted, it looks like you do not have this variable defined. Figure out where your nvim configuration is located (it will contain init.lua, probably) and create an ftplugin folder there (if it doesn't already exist), then follow the rest of the steps.

**Reply by xaviduds** [reply]
![screenshot](https://github.com/user-attachments/assets/8e4edf0a-4541-4441-917a-bd740d9976fa)
still no formatting, I feel like there is something really obvious I'm not doing

**Reply by jrebs** [reply]
What about if you just type 'jq' in the terminal.

**Reply by xaviduds** [reply]
$ jq
jq - commandline JSON processor [version 1.7.1]

Usage:  /run/current-system/sw/bin/jq [options] <jq filter> [file...]
        /run/current-system/sw/bin/jq [options] --args <jq filter> [strings...]
        /run/current-system/sw/bin/jq [options] --jsonargs <jq filter> [JSON_TEXTS...]

jq is a tool for processing JSON inputs, applying the given filter to
its JSON text inputs and producing the filter's results as JSON on
standard output.

The simplest filter is ., which copies jq's input to its output
unmodified except for formatting. For more advanced filters see
the jq(1) manpage ("man jq") and/or https://jqlang.github.io/jq/.

Example:

        $ echo '{"foo": 0}' | jq .
        {
          "foo": 0
        }

For listing the command options, use /run/current-system/sw/bin/jq --help.

~
❯ echo '{"foo": 0}' | jq .
{
  "foo": 0
}

**Reply by jrebs** [reply]
OK, so you do have jq installed. Are you sure that your options.lua is loading the settings you put above?

If you open nvim and load a .json file, and then type: ":set formatprg?" what output do you get in the statusline? I get "formatprg=jq" which means it is using jq as the formatter for json files.

**Reply by xaviduds** [reply]
![screenshot](https://github.com/user-attachments/assets/0e776f7e-b2c8-4e02-8ad9-6e36cdc7a003)
